### PR TITLE
NIFI-6398 Added the 'replace first' and 'replace all' strategy to ReplaceText

### PR DESCRIPTION
**Pull request by HorizonNet**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Adds a 'replace first' and 'replace all' strategy to the ReplaceText processor.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by markap14** [comment]
@HorizonNet can you explain the difference between "Replace All" and "Regex"? I believe they are intended to accomplish the same thing, but the existing Regex allows the user to change the behavior between Line-by-Line vs. Entire Text. Using the Line-by-Line mode is preferred if the Regex does not span multiple lines because it uses dramatically less heap. It also allows for back references, etc. Not sure if the new "Replace All" provides a capability that isn't currently supported, that I'm just missing?

The Replace First is a nice addition. We should ensure, though, any time that we create a `String` from `byte[]` that we pass in the character set in the constructor. It looks like it's used when serializing the `String` back to the `byte[]` but not when creating the `String` to begin with.

Finally, rather than using `IOUtils.toByteArray()`, would recommend creating an `byte[]` and then using `StringUtils.fillBuffer`, as is done in the `RegexReplace` strategy. Because we already know the size of the FlowFile, this is far more efficient (because it doesn't have to keep filling a buffer, creating a new one, and copying bytes over) and can cut down the amount of heap used to store the buffered data by up to 50% (because the ByteArrayOutputStream doubles the buffer size every time it runs out of space, so if it's already 512 KB and we need one more byte it creates a 1 MB buffer just to use 512 KB + 1 byte, for example, whereas a direct allocation takes exactly the right number of bytes).

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ReplaceText.java [hunk 1]
```diff
@@ -673,6 +689,63 @@ public boolean isAllDataBufferedForEntireText() {
...
+     * either be done line by line, meaning that the first occurrence within a line gets replaced, or for the entire
+     * text, where only the very first occurrence is replaced.
+     */
+    private class ReplaceFirst implements ReplacementStrategyExecutor {
+        @Override
+        public FlowFile replace(FlowFile flowFile, ProcessSession session, ProcessContext context, String evaluateMode,
+                Charset charset, int maxBufferSize) {
+            final String replacementValue = context.getProperty(REPLACEMENT_VALUE)
+                    .evaluateAttributeExpressions(flowFile).getValue();
+            final AttributeValueDecorator quotedAttributeDecorator = Pattern::quote;
```
Comment (ottobackwards) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ReplaceText.java]: According to the java documentation, you should quote the replacement value as well.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ReplaceText.java [hunk 2]
```diff
@@ -673,6 +689,63 @@ public boolean isAllDataBufferedForEntireText() {
...
+    /**
+     * Replaces all occurrences of a search value within a flow file with the given replacement value.
+     * The evaluation mode doesn't matter, because the result would be the same for both line-by-line or entire text.
+     */
+    private class ReplaceAll implements ReplacementStrategyExecutor {
+        @Override
+        public FlowFile replace(FlowFile flowFile, ProcessSession session, ProcessContext context, String evaluateMode,
+                Charset charset, int maxBufferSize) {
+            final String replacementValue = context.getProperty(REPLACEMENT_VALUE)
+                    .evaluateAttributeExpressions(flowFile).getValue();
```
Comment (ottobackwards) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ReplaceText.java]: According to the java documentation, you may need to quote the replacement text as well.

**Comment by joewitt** [comment]
closing due to inactivity

**Pull request closed without merging**